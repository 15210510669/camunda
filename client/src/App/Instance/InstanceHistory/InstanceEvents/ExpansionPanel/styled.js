import styled, {css} from 'styled-components';

import {themed, themeStyle, Colors} from 'modules/theme';
import {Down, Right} from 'modules/components/Icon';

export const ExpansionPanel = themed(styled.div`
  color: ${themeStyle({
    dark: '#ffffff',
    light: Colors.uiDark04
  })};
`);

const iconStyle = css`
  width: 16px;
  height: 16px;
  object-fit: contain;
  opacity: ${themeStyle({
    dark: 0.9
  })};
  color: ${themeStyle({
    dark: '#ffffff',
    light: Colors.uiDark04
  })};
`;

export const ExpandButton = styled.button`
  cursor: pointer;
  background: transparent;
  border: none;

  padding: 0;
  margin-right: 6px;
  top: 1px;
`;

export const DownIcon = themed(styled(Down)`
  ${iconStyle};
`);

export const RightIcon = themed(styled(Right)`
  ${iconStyle};
`);

export const Summary = themed(styled.div`
  display: flex;
  width: 100%;
  position: relative;
  padding-top: 7px;
  padding-bottom: 7px;

  font-weight: ${({bold}) => (!bold ? 'normal' : 'bold')};
`);

export const Details = styled.div`
  padding-left: 22px;
  display: ${({expanded}) => (expanded ? 'flex' : 'none')};
  flex-direction: column;
`;
