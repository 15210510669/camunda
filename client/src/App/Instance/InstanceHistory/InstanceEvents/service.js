/**
 * @returns {Array} of both:
 *  (1) events that don't have any activityInstanceId
 *  (2) events grouped by activityInstanceId
 * [
 *  // (1)
 *  {...event},
 *
 *  // (2)
 *  {
 *    name,
 *    events: [{...event}]
 *  }
 * ]
 */
export function getGroupedEvents({events, activitiesDetails}) {
  let groupedEvents = [];
  // make a deep clone of the activitiesDetails object
  let activitiesEvents = JSON.parse(JSON.stringify(activitiesDetails));

  events.forEach(event => {
    // if it doesn't have an activityInstanceId it doesn't belong to an activity events group
    if (!event.activityInstanceId) {
      return groupedEvents.push(event);
    }

    const targetActivityInstance = activitiesEvents[event.activityInstanceId];

    if (!targetActivityInstance.events) {
      targetActivityInstance.events = [event];
      // this allows to only push once a reference to the targetActivityInstance object to groupedEvents
      return groupedEvents.push(targetActivityInstance);
    }

    // modifying the targetActivityInstance object will also affect the one in the array since it's a reference
    targetActivityInstance.events.push(event);
  });

  return groupedEvents;
}
