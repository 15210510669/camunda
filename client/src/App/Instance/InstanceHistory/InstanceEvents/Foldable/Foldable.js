import React, {Children, cloneElement} from 'react';
import PropTypes from 'prop-types';

import * as Styled from './styled';

export default class Foldable extends React.Component {
  static propTypes = {
    children: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.node),
      PropTypes.node
    ])
  };

  state = {
    isFolded: true
  };

  toggleFold = () => {
    const {isFolded} = this.state;
    this.setState({isFolded: !isFolded});
  };

  render() {
    const children = Children.map(this.props.children, child =>
      cloneElement(child, {
        isFolded: this.state.isFolded,
        toggleFold: this.toggleFold
      })
    );

    return <Styled.Foldable>{children}</Styled.Foldable>;
  }
}

Foldable.Summary = function Summary({
  toggleFold,
  isFoldable,
  isFolded,
  children,
  ...props
}) {
  return (
    <Styled.Summary {...props}>
      {!isFoldable ? null : (
        <Styled.FoldButton onClick={toggleFold}>
          {!isFolded ? <Styled.DownIcon /> : <Styled.RightIcon />}
        </Styled.FoldButton>
      )}
      <Styled.SummaryLabel isFoldable={isFoldable}>
        {children}
      </Styled.SummaryLabel>
    </Styled.Summary>
  );
};

Foldable.Summary.propTypes = {
  toggleFold: PropTypes.func,
  isFoldable: PropTypes.bool,
  isFolded: PropTypes.bool,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]),
  isBold: PropTypes.bool
};

Foldable.Summary.defaultProps = {
  isFoldable: true
};

Foldable.Details = function Details(props) {
  return <Styled.Details {...props} />;
};
