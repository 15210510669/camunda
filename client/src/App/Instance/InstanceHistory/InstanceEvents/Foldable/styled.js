import styled, {css} from 'styled-components';

import {themed, themeStyle, Colors} from 'modules/theme';
import {ReactComponent as Down} from 'modules/components/Icon/down.svg';
import {ReactComponent as Right} from 'modules/components/Icon/right.svg';
import withStrippedProps from 'modules/utils/withStrippedProps';

const iconStyle = css`
  width: 16px;
  height: 16px;
  object-fit: contain;
  color: ${({theme, isSelected}) => {
    return isSelected || theme === 'dark'
      ? 'rgba(255, 255, 255, 0.9)'
      : Colors.uiDark04;
  }};
`;

export const FoldButton = styled.button`
  background: transparent;
  border: none;

  padding: 0;
  top: 1px;

  position: absolute;
  left: ${({indentation}) => `${indentation * 22 + 15}px`};
  top: 8px;
  z-index: 2;
`;

export const DownIcon = themed(styled(withStrippedProps(['isSelected'])(Down))`
  ${iconStyle};
`);

export const RightIcon = themed(styled(
  withStrippedProps(['isSelected'])(Right)
)`
  ${iconStyle};
`);

const borderBottomStyle = css`
1px solid ${themeStyle({
  dark: Colors.uiDark04,
  light: Colors.uiLight05
})}`;

export const Summary = themed(styled.div`
  position: relative;
  height: 32px;
  flex: 1;
  border-bottom: ${props => (!props.isFolded ? 'none' : borderBottomStyle)};
  border-color: ${props => (!props.isSelected ? '' : Colors.selections)};
`);

export const SummaryLabel = themed(styled.button`
  padding-left: ${({indentation}) => `${indentation * 22 + 37}px`};
  font-weight: ${({isFoldable, isSelected}) =>
    isSelected || isFoldable ? 'bold' : 'normal'};
  position: absolute;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  margin: 0;
  border: none;
  font-size: 14px;
  text-align: left;
  background-color: ${props =>
    !props.isSelected ? 'transparent' : Colors.selections};
  color: ${({theme, isSelected}) => {
    return isSelected || theme === 'dark'
      ? 'rgba(255, 255, 255, 0.9)'
      : Colors.uiDark04;
  }};
`);

export const Details = themed(styled.div`
  display: ${({isFolded}) => (!isFolded ? 'block' : 'none')};
  border-bottom: ${borderBottomStyle};
`);
