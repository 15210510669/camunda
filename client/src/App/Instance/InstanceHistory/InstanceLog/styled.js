import styled, {css} from 'styled-components';

import {Colors, themed, themeStyle} from 'modules/theme';
import {Document} from 'modules/components/Icon';
import BasicFlowNodeIcon from 'modules/components/FlowNodeIcon';
import {ACTIVITY_STATE} from 'modules/constants';
import withStrippedProps from 'modules/utils/withStrippedProps';

export const InstanceLog = themed(styled.div`
  flex: 1;
  display: flex;
  flex-direction: column;

  overflow: auto;
  border: solid 1px
    ${themeStyle({
      dark: Colors.uiDark04,
      light: Colors.uiLight05
    })};
  border-top: none;
  border-bottom: none;
`);

const backgroundColorStyle = css`
  &:nth-child(odd) {
    background-color: ${({theme, isSelected}) => {
      if (isSelected) {
        return Colors.selections;
      } else if (theme === 'dark') {
        return Colors.uiDark02;
      } else {
        return Colors.lightItemOdd;
      }
    }};
  }

  &:nth-child(even) {
    background-color: ${({theme, isSelected}) => {
      if (isSelected) {
        return Colors.selections;
      } else if (theme === 'dark') {
        return Colors.darkItemEven;
      } else {
        return Colors.lightItemEven;
      }
    }};
  }
`;

const colorStyle = css`
  color: ${({theme, isSelected, state}) => {
    if (state === ACTIVITY_STATE.INCIDENT) {
      return;
    } else {
      return isSelected || theme === 'dark'
        ? 'rgba(255, 255, 255, 0.9)'
        : Colors.uiDark04;
    }
  }};
`;

const iconPositionStyle = css`
  position: relative;
  top: 3px;
`;
export const LogEntry = themed(styled.button`
  background: transparent;
  margin: 0;
  border: none;
  padding: 5px;
  padding-left: 43px;
  height: 32px;
  text-align: left;

  ${backgroundColorStyle};
  ${colorStyle};

  cursor: pointer;

  font-size: 14px;
  font-weight: ${({isSelected}) => (!!isSelected ? 'bold' : 'normal')};
`);

export const Header = themed(styled(LogEntry)`
  padding-left: 23px;
  color: ${themeStyle({
    dark: 'rgba(255, 255, 255, 0.9)',
    light: 'rgba(69, 70, 78, 0.9)'
  })};
  font-weight: bold;
`);

export const DocumentIcon = themed(styled(
  withStrippedProps(['isSelected'])(Document)
)`
  width: 16px;
  height: 16px;
  margin-right: 8px;

  ${colorStyle};
  ${iconPositionStyle};
`);

export const FlowNodeIcon = themed(styled(
  withStrippedProps(['isSelected'])(BasicFlowNodeIcon)
)`
  margin-right: 8px;

  ${colorStyle};
  ${iconPositionStyle};
`);
