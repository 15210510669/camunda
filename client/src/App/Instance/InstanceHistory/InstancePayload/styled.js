import styled from 'styled-components';
import {Colors, themed, themeStyle} from 'modules/theme';

export const InstancePayload = themed(styled.div`
  flex: 1;
  font-size: 14px;

  overflow: auto;

  position: relative;
  background-color: ${themeStyle({
    dark: Colors.uiDark02,
    light: Colors.uiLight04
  })};

  color: ${themeStyle({
    dark: 'rgba(255, 255, 255, 0.8)',
    light: 'rgba(98, 98, 110, 0.8)'
  })};
`);

export const PayloadContent = styled.div`
  position: absolute;
  height: 100%;
  width: 100%;
  display: flex;
  align-items: center;
`;

export const Placeholder = themed(styled.span`
  text-align: center;
  top: 0;
  left: 0;
  width: 100%;
  font-size: 16px;
  color: ${themeStyle({
    dark: '#dedede',
    light: Colors.uiLight06
  })};
`);

export const Table = themed(styled.table`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  border-spacing: 0;
  border-collapse: collapse;
`);

export const TH = themed(styled.th`
  font-style: italic;
  font-weight: normal;
  text-align: left;
  padding-left: 17px;
  height: 31px;
`);

export const TD = themed(styled.td`
  color: ${themeStyle({
    dark: 'rgba(255, 255, 255, 0.9)',
    light: 'rgba(98, 98, 110, 0.9)'
  })};
  font-weight: ${props => (props.isBold ? 'bold' : 'normal')};
  padding-left: 17px;
  height: 32px;
`);

export const TR = themed(styled.tr`
  border-width: 1px 0;
  border-style: solid;
  border-color: ${themeStyle({
    dark: Colors.uiDark04,
    light: Colors.uiLight05
  })};

  &:first-child {
    border-top: none;
  }
`);
