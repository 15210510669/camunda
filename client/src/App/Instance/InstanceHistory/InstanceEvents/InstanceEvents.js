import React from 'react';
import PropTypes from 'prop-types';

import {isEmpty} from 'modules/utils';
import {ACTIVITY_STATE, EVENT_TYPE, EVENT_SOURCE_TYPE} from 'modules/constants';

import Foldable from './Foldable';
import * as Styled from './styled';

export default class InstanceEvents extends React.Component {
  static propTypes = {
    groupedEvents: PropTypes.array
  };

  /**
   * Filters foldable details and renders them.
   * @param {object} details: An object of key-value entries where the value can be a string | number | object | null.
   * e.g. {
   *  jobCustomHeaders: {}
      jobDeadline: "2018-08-31T07:16:52.805+0000"
      jobId: "8590888008"
      jobRetries: 3
   * }
   */
  renderFoldableDetails = details => {
    return (
      Object.entries(details)
        // filter out empty primitives and objects
        .filter(([_, value]) => {
          if (typeof value === 'object') {
            return !isEmpty(value);
          }

          return Boolean(value);
        })
        .map(([key, value], idx) => {
          // if the value is a primitive, render a DataEntry
          if (typeof value !== 'object') {
            return (
              <Styled.DataEntry
                key={idx}
              >{`${key}: ${value}`}</Styled.DataEntry>
            );
          }

          // if the value is of type object, render a Foldable
          return (
            <Foldable key={idx}>
              <Foldable.Summary>{key}</Foldable.Summary>
              <Foldable.Details>
                {this.renderFoldableDetails(value)}
              </Foldable.Details>
            </Foldable>
          );
        })
    );
  };

  renderEvent = ({eventType, eventSourceType, metadata}, idx) => {
    const key = `${eventType}${idx}`;
    const isOpenIncidentEvent =
      eventType === EVENT_TYPE.CREATED &&
      eventSourceType === EVENT_SOURCE_TYPE.INCIDENT;

    return (
      <Foldable key={key}>
        <Styled.EventFoldableSummary
          isFoldable={!isEmpty(metadata)}
          isOpenIncidentEvent={isOpenIncidentEvent}
        >
          {!isOpenIncidentEvent ? '' : <Styled.IncidentIcon />}
          {eventType}
        </Styled.EventFoldableSummary>
        <Foldable.Details>
          {!!metadata && this.renderFoldableDetails(metadata)}
        </Foldable.Details>
      </Foldable>
    );
  };

  renderEventsGroup = ({name, events, state}, idx) => {
    const key = `${name}${idx}`;

    return (
      <Foldable key={key}>
        <Styled.GroupFoldableSummary>
          {name}
          {state !== ACTIVITY_STATE.INCIDENT ? '' : <Styled.IncidentIcon />}
        </Styled.GroupFoldableSummary>
        <Foldable.Details>{events.map(this.renderEvent)}</Foldable.Details>
      </Foldable>
    );
  };

  renderGroup = (group, idx) => {
    return (
      <Styled.EventEntry key={idx}>
        {group.events
          ? this.renderEventsGroup(group, idx)
          : this.renderEvent(group, idx)}
      </Styled.EventEntry>
    );
  };

  render() {
    return (
      <Styled.InstanceEvents {...this.props}>
        <Styled.EventsContainer>
          {this.props.groupedEvents
            ? this.props.groupedEvents.map(this.renderGroup)
            : null}
        </Styled.EventsContainer>
      </Styled.InstanceEvents>
    );
  }
}
