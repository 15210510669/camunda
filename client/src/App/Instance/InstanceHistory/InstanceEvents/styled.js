import styled from 'styled-components';

import {Colors, themed, themeStyle} from 'modules/theme';

import Foldable from './Foldable';

export const InstanceEvents = themed(styled.div`
  flex: 1;
  position: relative;

  overflow: auto;
  border: solid 1px
    ${themeStyle({
      dark: Colors.uiDark04,
      light: Colors.uiLight05
    })};
  border-top: none;
  border-bottom: none;
`);

export const EventsContainer = styled.div`
  position: absolute;
  height: 100%;
  width: 100%;
  display: flex;
  flex-direction: column;
  min-height: min-content;
`;

export const EventEntry = themed(styled.div`
  padding-left: 15px;
  border-bottom: 1px solid
    ${themeStyle({
      dark: Colors.uiDark04,
      light: Colors.uiLight05
    })};
`);

export const DataEntry = themed(styled.div`
  opacity: 0.9;
  color: ${themeStyle({
    light: Colors.uiLight06
  })};
  padding-top: 7px;
  padding-bottom: 7px;
`);

export const GroupFoldableSummary = styled(Foldable.Summary)`
  font-weight: bold;
`;
