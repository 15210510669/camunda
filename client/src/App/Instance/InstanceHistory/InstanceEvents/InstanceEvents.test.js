import React from 'react';
import {shallow} from 'enzyme';

import * as api from 'modules/api/events/events';
import {mockResolvedAsyncFn, flushPromises} from 'modules/testUtils';

import InstanceEvents from './InstanceEvents';
import * as Styled from './styled';
import Foldable from './Foldable';

const fooActivityEvents = [
  {
    eventType: 'fooCreated',
    activityInstanceId: 'foo',
    metadata: {
      a: 'b'
    }
  },
  {
    eventType: 'fooActiviated',
    activityInstanceId: 'foo'
  }
];

const barActivityEvents = [
  {
    eventType: 'barCreated',
    activityInstanceId: 'bar'
  },
  {
    eventType: 'barActivated',
    activityInstanceId: 'bar'
  }
];

const instanceEvents = [
  {
    eventType: 'baz',
    metadata: {
      c: {
        d: 'e',
        e: null
      },
      f: null
    }
  }
];

const mockEvents = [
  ...fooActivityEvents,
  ...barActivityEvents,
  ...instanceEvents
];

api.fetchEvents = mockResolvedAsyncFn(mockEvents);

const mockInstance = {id: 'foo'};

const fooActivityDetails = {name: 'foo name'};
const barActivityDetails = {name: 'bar name'};
const mockActivitiesDetails = {
  foo: fooActivityDetails,
  bar: barActivityDetails
};

describe('InstanceEvents', () => {
  beforeEach(() => {
    api.fetchEvents.mockClear();
  });

  it('should render empty events container by default', () => {
    // given
    const node = shallow(<InstanceEvents instance={mockInstance} />);

    // then
    expect(node.state()).toEqual({
      events: null,
      groupedEvents: null
    });
    const EventsContainerNode = node.find(Styled.EventsContainer);
    expect(EventsContainerNode).toHaveLength(1);
    expect(EventsContainerNode.children().length).toBe(0);
  });

  it("should not render grouped events if events or activities didn't change", async () => {
    // given
    const node = shallow(<InstanceEvents instance={mockInstance} />);
    await flushPromises();
    node.update();

    // then
    expect(node.state()).toEqual({
      events: mockEvents,
      groupedEvents: null
    });
    const EventsContainerNode = node.find(Styled.EventsContainer);
    expect(EventsContainerNode).toHaveLength(1);
    expect(EventsContainerNode.children().length).toBe(0);
  });

  it('should render grouped events if events or activities changed', async () => {
    // given
    const expectedGroupedEvents = [
      {
        events: fooActivityEvents,
        name: fooActivityDetails.name
      },
      {
        events: barActivityEvents,
        name: barActivityDetails.name
      },
      ...instanceEvents
    ];
    const node = shallow(<InstanceEvents instance={mockInstance} />);
    await flushPromises();
    node.setProps({activitiesDetails: mockActivitiesDetails});
    node.update();

    // then
    expect(node.state().events).toEqual(mockEvents);
    expect(node.state('groupedEvents')).toEqual(expectedGroupedEvents);
    // Foldable for each event and each group
    const FoldableNodes = node.find(Foldable);
    expect(FoldableNodes).toHaveLength(8);

    // Foo Foldable
    const FooFoldableNode = FoldableNodes.at(0);
    // Foo Summary
    const SummaryNode = FooFoldableNode.find(Foldable.Summary).at(0);
    expect(SummaryNode.prop('bold')).toBe(true);
    expect(SummaryNode.contains(fooActivityDetails.name));

    // Foo Details
    const FoldableDetailsNode = FooFoldableNode.find(Foldable.Details).at(0);
    expect(FoldableDetailsNode).toHaveLength(1);

    // Foo Events Foldables
    expect(FoldableDetailsNode.find(Foldable)).toHaveLength(
      fooActivityEvents.length
    );

    // Instance Events Foldables
    const InstanceFoldableNode = FoldableNodes.at(6);
    // Instance Event Summary
    const InstanceEventSummaryNode = InstanceFoldableNode.find(
      Foldable.Summary
    ).at(0);
    expect(InstanceEventSummaryNode.prop('bold')).not.toBe(true);
    expect(InstanceEventSummaryNode.contains(instanceEvents[0].eventType));

    // Instance Events Foldables
    expect(InstanceFoldableNode.find(Foldable)).toHaveLength(2);

    expect(node).toMatchSnapshot();
  });
});
