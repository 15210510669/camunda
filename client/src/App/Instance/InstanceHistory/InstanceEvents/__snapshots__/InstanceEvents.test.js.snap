// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstanceEvents should render grouped events if events or activities changed 1`] = `
<Themed(styled.div)
  activitiesDetails={
    Object {
      "bar": Object {
        "name": "bar name",
      },
      "foo": Object {
        "name": "foo name",
      },
    }
  }
  instance={
    Object {
      "id": "foo",
    }
  }
>
  <styled.div>
    <Themed(styled.div)
      key="0"
    >
      <Foldable
        key="foo name0"
      >
        <Summary
          isBold={true}
          isFoldable={true}
        >
          foo name
        </Summary>
        <Details>
          <Foldable
            key="fooCreated0"
          >
            <Summary
              isFoldable={true}
            >
              fooCreated
            </Summary>
            <Details>
              <Themed(styled.div)
                key="0"
              >
                a: b
              </Themed(styled.div)>
            </Details>
          </Foldable>
          <Foldable
            key="fooActiviated1"
          >
            <Summary
              isFoldable={false}
            >
              fooActiviated
            </Summary>
            <Details />
          </Foldable>
        </Details>
      </Foldable>
    </Themed(styled.div)>
    <Themed(styled.div)
      key="1"
    >
      <Foldable
        key="bar name1"
      >
        <Summary
          isBold={true}
          isFoldable={true}
        >
          bar name
        </Summary>
        <Details>
          <Foldable
            key="barCreated0"
          >
            <Summary
              isFoldable={false}
            >
              barCreated
            </Summary>
            <Details />
          </Foldable>
          <Foldable
            key="barActivated1"
          >
            <Summary
              isFoldable={false}
            >
              barActivated
            </Summary>
            <Details />
          </Foldable>
        </Details>
      </Foldable>
    </Themed(styled.div)>
    <Themed(styled.div)
      key="2"
    >
      <Foldable
        key="baz2"
      >
        <Summary
          isFoldable={true}
        >
          baz
        </Summary>
        <Details>
          <Foldable
            key="0"
          >
            <Summary
              isFoldable={true}
            >
              c
            </Summary>
            <Details>
              <Themed(styled.div)
                key="0"
              >
                d: e
              </Themed(styled.div)>
            </Details>
          </Foldable>
        </Details>
      </Foldable>
    </Themed(styled.div)>
  </styled.div>
</Themed(styled.div)>
`;
